dali/pipeline/operators/decoder/nvjpeg_decoder_decoupled_api.h

```diff
@@ -0,0 +1,359 @@
+  explicit nvJPEGDecoder(const OpSpec& spec) :
+    Operator<MixedBackend>(spec),
+    output_image_type_(spec.GetArgument<DALIImageType>("output_type")),
+    hybrid_huffman_threshold_(spec.GetArgument<unsigned int>("hybrid_huffman_threshold")),
+    output_info_(batch_size_),
+    image_decoders_(batch_size_),
+    image_states_(batch_size_),
+    decode_params_(batch_size_),
+    decoder_host_state_(batch_size_),
+    decoder_huff_hybrid_state_(batch_size_),
+    output_shape_(batch_size_),
+    jpeg_streams_(num_threads_ * 2),
+    pinned_buffer_(num_threads_ * 2),
+    curr_pinned_buff_(num_threads_, 0),
+    device_buffer_(num_threads_),
+    streams_(num_threads_),
+    events_(num_threads_),
+    thread_pool_(num_threads_,
+                 spec.GetArgument<int>("device_id"),
+                 true /* pin threads */) {
+    NVJPEG_CALL(nvjpegCreateSimple(&handle_));
+
+    size_t device_memory_padding = spec.GetArgument<Index>("device_memory_padding");
+    size_t host_memory_padding = spec.GetArgument<Index>("host_memory_padding");
+    NVJPEG_CALL(nvjpegSetDeviceMemoryPadding(device_memory_padding, handle_));
+    NVJPEG_CALL(nvjpegSetPinnedMemoryPadding(host_memory_padding, handle_));
+
+    // to create also in GPU Op
+    NVJPEG_CALL(nvjpegDecoderCreate(
+                      handle_, NVJPEG_BACKEND_HYBRID, &decoder_huff_host_));
+    NVJPEG_CALL(nvjpegDecoderCreate(
+                      handle_, NVJPEG_BACKEND_GPU_HYBRID, &decoder_huff_hybrid_));
+
+
+    for (int i = 0; i < batch_size_; i++) {
+      NVJPEG_CALL(nvjpegDecodeParamsCreate(handle_, &decode_params_[i]));
+      NVJPEG_CALL(nvjpegDecodeParamsSetOutputFormat(decode_params_[i],
+                                                    GetFormat(output_image_type_)));
+      NVJPEG_CALL(nvjpegDecodeParamsSetAllowCMYK(decode_params_[i], true));
+
+      // We want to use nvJPEG default pinned allocator
+
+      NVJPEG_CALL(nvjpegDecoderStateCreate(handle_,
+                                        decoder_huff_host_,
+                                        &decoder_host_state_[i]));
+      NVJPEG_CALL(nvjpegDecoderStateCreate(handle_,
+                                        decoder_huff_hybrid_,
+                                        &decoder_huff_hybrid_state_[i]));
+    }
+
+    // GPU
+    // create the handles, streams and events we'll use
+    // We want to use nvJPEG default device allocator
+    for (int i = 0; i < num_threads_ * 2; ++i) {
+      NVJPEG_CALL(nvjpegJpegStreamCreate(handle_, &jpeg_streams_[i]));
+      NVJPEG_CALL(nvjpegBufferPinnedCreate(handle_, nullptr, &pinned_buffer_[i]));
+    }
+    for (int i = 0; i < num_threads_; ++i) {
+      NVJPEG_CALL(nvjpegBufferDeviceCreate(handle_, nullptr, &device_buffer_[i]));
+      CUDA_CALL(cudaStreamCreateWithFlags(&streams_[i], cudaStreamNonBlocking));
+      CUDA_CALL(cudaEventCreate(&events_[i]));
+    }
+    CUDA_CALL(cudaEventCreate(&master_event_));
+  }
+
+  virtual ~nvJPEGDecoder() noexcept(false) {
```

Comment:
It's bad idea to throw something from dtor, while all CUDA_ ans NVJPEG_ calls do it. I'd change it to `noexcept`, since if any of this calls throw, it's not a situation we can recover anyway...

Reply:
That's a well known issue (Janusz mentioned in another PR that he marked as `noexcept(false)` to make the linter happy. We should go through those destructors and fix it (that is, make sure they don't throw). That should not be in the scope of this PR, I believe

Reply:
I agree, I just copy pasted the dtor name from old nvJPEGDecoder tbh